bundles/org.openhab.binding.energidataservice/src/main/java/org/openhab/binding/energidataservice/internal/handler/EnergiDataServiceHandler.java

```diff
@@ -321,19 +322,27 @@ private void updateCurrentSpotPrice() {
             return;
         }
         BigDecimal spotPrice = cacheManager.getSpotPrice();
-        updateState(CHANNEL_SPOT_PRICE,
-                spotPrice != null ? getEnergyPrice(spotPrice, config.getCurrency()) : UnDefType.UNDEF);
+        updatePriceState(CHANNEL_SPOT_PRICE, spotPrice, config.getCurrency());
     }
 
     private void updateCurrentTariff(String channelId, @Nullable BigDecimal tariff) {
         if (!isLinked(channelId)) {
             return;
         }
-        updateState(channelId, tariff != null ? getEnergyPrice(tariff, CURRENCY_DKK) : UnDefType.UNDEF);
+        updatePriceState(channelId, tariff, CURRENCY_DKK);
     }
 
-    private QuantityType<EnergyPrice> getEnergyPrice(BigDecimal price, Currency currency) {
-        return new QuantityType<>(price + " " + currency.getSymbol() + "/kWh");
+    private void updatePriceState(String channelID, @Nullable BigDecimal price, Currency currency) {
+        updateState(channelID, price != null ? getEnergyPrice(price, currency) : UnDefType.UNDEF);
+    }
+
+    private State getEnergyPrice(BigDecimal price, Currency currency) {
+        try {
+            return new QuantityType<>(price + " " + currency.getSymbol() + "/kWh");
+        } catch (IllegalArgumentException e) {
+            logger.trace("Unable to create QuantityType, falling back to DecimalType", e);
+            return new DecimalType(price);
```

Comment:
Thanks. I'm now verifying the unit before trying to instantiate `QuantityType` from it.

Additionally I will now fallback to currency code if there is no symbol for the currency, although my tests suggest that currency code as unit is not supported. Also I'm now using `getSymbol()` on the returned `Unit` rather than [Currency.getSymbol()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Currency.html#getSymbol()), because that is display unit for the `Locale` and thus prone to failure (e.g. "$" vs. "US$").

I don't think using base currency is such a good idea, because that would mean explicitly providing a wrong unit. The end result is currently the same though.

Reply:
I see this only as a work-around to ensure reliable state updates in 4.1, would you agree? And do you see anything in all this that is worth creating an issue for in core?

We currently have the limitation that the UoM system know only one currency. When this will change, I would somehow expect:
- Always being able to update state for known currencies, even if there is no currency exchange service provided/installed - without risking `IllegalArgumentException` being thrown.
- In this case IMHO it would still be better to keep the source currency unit rather than overwriting by user configured currency unit without converting the amount. I know this contradicts 4.0 UoM because we cannot enforce "target unit", and I don't have a good proposal for tackling that. It's different, because we can always expect to be able to convert °C to °F, but we can't always expect to be able to convert DKK to EUR. And we cannot make the assumption that exchange rate is 100, because then prices will be plain misleading.
- Being able to update state with currency code as unit rather than symbol. When all currencies will become available and we are able to convert between them, symbols can become ambiguous as previously discussed.